xmodule/services.py

```diff
@@ -308,3 +312,133 @@ def _handle_deprecated_progress_event(self, block, event):
         # in order to avoid duplicate work and possibly conflicting semantics.
         if not getattr(block, 'has_custom_completion', False):
             self.completion_service.submit_completion(block.scope_ids.usage_id, 1.0)
+
+
+class ProblemFeedbackService(Service):
+    """
+    An XBlock Service that allows XModules to define:
+
+    - if the answers are visible
+    - whether correctness is currently hidden. When correctness is hidden, this
+    limits the user's access to the correct/incorrect flags, messages, problem
+    scores, and aggregate subsection and course grades.
+    """
+    def __init__(self, block=None, user_is_staff=False, **kwargs):
+        super().__init__(**kwargs)
+        self._user_is_staff = user_is_staff
+        # This is needed because the `Service` class initialization expects the XBlock passed as an `xblock` keyword
+        #  argument, but the `service` method from the `DescriptorSystem` passes a `block`.
+        self._xblock = self.xblock() or block
+
+    def is_past_due(self):
+        """
+        Returns if the problem is past its due date.
+        """
+        if self._xblock:
+            due_date = getattr(self._xblock, 'close_date', None) or getattr(self._xblock, 'due', None)
+            return (due_date is None or datetime.now(UTC) > due_date)
```

Comment:
This is related to the [comment I left above](https://github.com/openedx/edx-platform/pull/34320#issuecomment-1976850180).

One on hand, we want it to be `false` when there is no due date set (I'll ignore `close_date`, because it's `due + graceperiod`, so if `due` is `None`, `close_date` is also `None`), for example when we want the learner to be able to complete problems in the self paced course where there are no due dates. On another hand, there is logic related to displaying the answer and it's correctness, that depends on the due date, and in the scenario when there is no due date, we want the learner to be able to see the answer right after they've completed the problem, and in that case we want to it to return `true`. So these to interpretations of the `is_past_due` are conflicting. We could keep this method internal to this service.

Re `InheritanceMixin`, I tried removing the `is_past_due` method of the capa XBlock, and the tests started failing, so I decided to play it save, and keep it there, as I was trying to make the least number of changes to an already approved PR.

Reply:
Also, the reason I decided to check for both `close_date` and `due` properties of the XBlock, is because on the code block referenced below, XBlocks that are passed to the service seem to not have an implementation for `close_date`, because the tests start failing. That's why I added the fallback to the `due` property.

https://github.com/openedx/edx-platform/blob/5907b7e4f5ffa18dba4598871816b9fa353f8dcf/lms/djangoapps/grades/subsection_grade.py#L60-L64

I'm not sure why the don't inherit from the `InheritanceMixin`, but if they were, they would have the `close_date` property.

You can see failed test [here](https://github.com/openedx/edx-platform/actions/runs/8141411936/job/22248710573).

Reply:
I tried removing the `is_past_due` method of the `ProblemXBlock`, located the tests that are failing (bellow for reference), and found that the in the factory that creates the xblocks for the tests the line that should add these mixins is commented out. Do you know why it was commented out? And whether we can uncomment them

```
================================================================================================================================= short test summary info ==================================================================================================================================
FAILED xmodule/tests/test_capa_block.py::ProblemBlockTest::test_submit_problem_closed - AttributeError: 'ProblemBlock' object has no attribute 'is_past_due'
FAILED xmodule/tests/test_delay_between_attempts.py::XModuleQuizAttemptsDelayTest::test_still_cannot_submit_after_max_attempts - AttributeError: 'ProblemBlock' object has no attribute 'is_past_due'
```

https://github.com/openedx/edx-platform/blob/5907b7e4f5ffa18dba4598871816b9fa353f8dcf/xmodule/tests/__init__.py#L239-L243

Though uncommenting these lines doesn't fix the issue.

Reply:
@Cup0fCoffee, I pushed 184516500f, which initializes XBlock mixins in CAPA tests.

Short explanation:
1. `CapaFactory` creates a new ProblemBlock. As the first step, it initializes the runtime here:
   https://github.com/openedx/edx-platform/blob/184516500f/xmodule/tests/test_capa_block.py#L121-L125
   (Note: I renamed "system" to "runtime" because the "system" is [a deprecated term](https://github.com/openedx/edx-platform/blob/184516500f/openedx/core/djangoapps/xblock/runtime/shims.py#L321-L331).)
2. The `get_test_system` function initializes runs `get_test_descriptor_system` to create a new `TestDescriptorSystem`. As we can see below, this class already contains the `InheritanceMixin` during initialization:
   https://github.com/openedx/edx-platform/blob/184516500f/xmodule/tests/__init__.py#L259
3. The `mixins` arg passed there is then used during the initialization of the `Runtime` class. It creates [a new `Mixologist` instance](https://github.com/openedx/XBlock/blob/bd289d2f9f17c62276df24d8a57b437b160114ec/xblock/runtime.py#L568).
4. However, the mixing (`mixologist.mix(cls)(...)`) needs to be explicitly executed. As it's implicitly executed in the [`construct_xblock_from_class` method](https://github.com/openedx/XBlock/blob/bd289d2f9f17c62276df24d8a57b437b160114ec/xblock/runtime.py#L641-L651), I decided to use it instead of initializing the block directly from the `ProblemBlock` class.

Reply:
@Agrendalath Nice! Thank you for the explanation. I found another place in xmodule tests that needs this change, and applied it.

I've also removed the `close_date` and `is_past_due` methods from the `ProblemXBlock`.